### PR TITLE
How to Build Authentication Api with Jwt Token in Nodejs

### LINKED ISSUE #2226: How to Build Authentication Api with Jwt Token in Nodejs
### Proposed title of the article
How to Build Open Authentication Api with Jwt Token and Nodejs

### Introduction
Introduction paragraph (2-3 paragraphs):

JSON Web Token (JWT) is an open standard (RFC 7519) that specifies a compact and self-contained method for securely transmitting information as a JSON object between two parties. Since it is digitally signed, this information can be checked and trusted. JWTs can be signed with a password (using the HMAC algorithm) or with an RSA or ECDSA public/private key pair.

Node.js is a cross-platform runtime environment for building server-side and networking applications that is open-source. Node.js programs are written in JavaScript and run on OS X, Microsoft Windows, and Linux using the Node.js runtime.

### Key takeaways:
By the end of this tutorial, the reader will:

- Learn JWT Basics and how to use them.
- Understand how JWT works behind the scene in NodeJS for Authentication
- Develop an API using JWT Token for Authentication in NodeJS.

### References:
Templates to use as guides
https://jwt.io/
https://nodejs.org/en/

### PR DESCRIPTION
**Pull request by Olanetsoft**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This Fixes #2226

### CODE REVIEW

**Comment by geoffrey45** [comment]
Hello @Olanetsoft 

Thanks for submitting your article for review. Please read and check the necessary check-boxes in the pull request header.

**Comment by Olanetsoft** [comment]
okay

**Comment by Olanetsoft** [comment]
@geoffrey45 updated

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Comment by Olanetsoft** [comment]
@hectorkambow thank you for the feedback, I'm patiently waiting.

File: content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+
+We will install mongoose because I will be using MongoDB for storage while we validate user credentials against what we have in our database. The whole authentication process is not limited to the database we'll be using in this article.
+
+```bash
+npm install mongoose express jsonwebtoken dotenv bcryptjs
+npm install nodemon -D
+```
+
+#### Step 5 - Create a NodeJS server and connect your database
+
```
Comment (geoffrey45) [inline content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md:75]: Provide a brief explanation of the code in this section.

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
Hello @Olanetsoft 

Thank you for submitting your article for review. It's very informative. Please provide an explanation of the code in step 5

File: content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md (hunk 1 above)
Comment (Olanetsoft) [reply content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md:75]: ok

**Comment by Olanetsoft** [comment]
@geoffrey45 updated

**Review by geoffrey45 — APPROVED** [review]
@Olanetsoft i was taking your article for a tech review and I encountered some issues. Please see the attached comments.

File: content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+- Linux
+
+```bash
+mkdir jwt-project
+cd jwt-project
+npm init -y
+```
+
+#### Step 3 - Create files and directories as shown below
+![folder structure](/engineering-education/how-to-build-authentication-api-with-jwt-token-in-nodejs/folder-struc.png)
```
[the remark this answers is not in the document]
Comment (Olanetsoft) [reply content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md:62]: done

File: content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+```
+
+#### Step 5 - Create a NodeJS server and connect your database
+
+Now let's create our NodeJS server and connect our database by adding the following snippet to you `app.js`, `index.js` `, database.js` `.env`,.
+
+In our `database.js.`
+
+```javascript
+const mongoose = require("mongoose");
```
[the remark this answers is not in the document]
Comment (Olanetsoft) [reply content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md:81]: done

File: content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+- encrypt user password.
+- create a user in our database.
+Finally, create a signed JWT token.
+
+```javascript
+app.post("/register", async (req, res) => {
+  try {
+    // Get user input
+    const { first_name, last_name, email, password } = req.body;
+
```
[the remark this answers is not in the document]
Comment (Olanetsoft) [reply content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md:227]: done

File: content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+![error](/engineering-education/how-to-build-authentication-api-with-jwt-token-in-nodejs/error-response.png)
+
+We can now add a `token` in the header with the key `x-access-token` and re-test. See the image below for the response.
+
+![success](/engineering-education/how-to-build-authentication-api-with-jwt-token-in-nodejs/success-response.png)
+
+Here is what our `app.js` file looks like:
+
+```Javascript
```
[the remark this answers is not in the document]
Comment (Olanetsoft) [reply content/articles/how-to-build-authentication-api-with-jwt-token-in-nodejs/index.md]: updated

**Comment by Olanetsoft** [comment]
@geoffrey45  All changes Updated

**Comment by Olanetsoft** [comment]
@geoffrey45  thanks for the minor edit.

**Review by geoffrey45 — APPROVED** [review]

**Comment by hectorkambow** [comment]
keep up the great work gents. @Olanetsoft 

cc @geoffrey45 

just see the minor changes I have made 🚀

**Pull request merged**